### ISSUE TITLE
Instance reference

### ISSUE THREAD

**Issue by Mouvedia** [ericelliott/rtype#110]
Is there an idiomatic way to refer to an instance when declaring a constructor function?
If not would it make sense to allow its omission? In that case it would _not_ default to `Void` though.

**Comment by ericelliott** [comment]
Can you give me an example?

**Comment by Mouvedia** [comment]
@ericelliott https://github.com/ericelliott/rtype/pull/109#issuecomment-246089573

**Comment by ericelliott** [comment]
Does `this` in #120 work?

**Comment by Mouvedia** [comment]
The purpose was to have something simple and straightforward without having to define the instance.  It would imply its nature and its origin, nothing more.

cf [constructors chapter](https://github.com/ericelliott/rtype#constructors)

#120 doesn't meet that requirement.

**Comment by ericelliott** [comment]
I mean just using the `this` reference as demonstrated in the #120 ticket.

```js
interface Foo {
  (n) => Foo[n],
  addTo (this: Foo[n], t) => Number
};
```

The implementation of Foo might look like this:

```js
const Foo = n => ({
  addTo: (t) => n + t
});

Foo(3).addTo({ valueOf: () => 3 }); // 6
```

In the above signature, both `n` and `t` are type variables, meaning they can represent any type. Technically, they represent any type belonging to the typeclass that requires a `valueOf()` method to be present, but that's a tangent.

This really starts to get useful when dealing with higher-order functions:

```
interface Identity {
  (a) => Functor[a],
  map (this: Functor[a], a => b) => Functor[b]
}
```

Which means, an instance of Identity is a Functor (mappable) containing a type `a`, takes a function from type `a` to type `b`, and returns a new functor of type `b`.

Adding `this` to the function signature in this context clarifies that the method is being applied to the value of `this` (i.e., using `this` as a parameter, rather than applying the function only to the arguments passed in).

We could probably explain and demonstrate `this` with a simpler example.

Does specifying `this` as the first argument solve the problem you need solved?

**Comment by Mouvedia** [comment]
> Does specifying `this` as the first argument solve the problem you need solved?

No because I think it doesn't belong there. I agree that it should be indicated* somewhere but not as the first parameter, that doesn't make sense to me.

Anyhow this issue is not about `this` but about instances returned by constructors. How to convey that they are instances without having to affix them with `Instance`. I think it is something that comes up frequently enough that it warrants a default behaviour.

\* #68

**Comment by ericelliott** [comment]
> No because I think it doesn't belong there. I agree that it should be indicated* somewhere but not as the first parameter, that doesn't make sense to me.

Why doesn't it make sense? A function that uses `this` is functionally equivalent to a function that was passed a value for it as an argument.

> Anyhow this issue is not about this but about instances returned by constructors.

Unless you change the default behavior and turn a constructor into a funky factory, constructors return `this` as the instance. Just set the constructor return type to whatever you like (which is already supported using the return type of the function signature).

IMO, it's not meaningful in JavaScript to say `foo` is an instance of `bar` in the same way that it's meaningful in Java, where a member of a class is statically typed to that class. No such relationship exists in JavaScript. JavaScript's `instanceof` operator lies, and the relationship it checks (identity comparisons in the prototype chain) is *dynamic*, and mutable, not *static* and immutable.

Because `instanceof` is not a meaningful, immutable, static property in JavaScript, I don't think it has any place in a JavaScript type notation system.

**Comment by Mouvedia** [comment]
> Just set the constructor return type to whatever you like (which is already supported using the return type of the function signature).

That would require to declare something beforehand. I was hoping to have something that can be omitted but the default is `Void` and I don't wanna complicate things further by having a special case for constructors. Still I find it odd that you have to declare it—inline or not—or even state that it's returning an object; that's common sense.

> Unless you change the default behavior and turn a constructor into a funky factory, constructors return `this` as the instance.

Would `new (options: Object) => this` make sense to you?
It would immediately make me wonder: which `this` is it referring to?

**Comment by ericelliott** [comment]
> That would require to declare something beforehand.

If you don't need to specify the shape of the return type, just use a type variable: `new User() => u`. Otherwise, you'll need to declare something for it to conform to... e.g., `new User() => UserInstance`.

> Would `new (options: Object) => this` make sense to you?

No.

**Comment by Mouvedia** [comment]
> Why doesn't it make sense?

That's just how I feel about it. `this` is not an argument to me, it's a context.
Most of the time if you are compelled to add it, you are trying to restrict the caller specificities.
That's useful, but rare; it should probably be relegated to a new keyword or at least be clearly separated.

**Comment by ericelliott** [comment]
How would you relegate it to the `new` keyword? What about factory functions that don't need the `new` keyword?

**Comment by Mouvedia** [comment]
> How would you relegate it to the new keyword?

You didn't understand what I said: "relegated to **a** new keyword".
I meant something akin to `throws` or `requires`. For example `binds`.
But that's off-topic on this issue.

This issue is about having a default that permit to omit the definition of an instance object.

**Comment by Mouvedia** [comment]
> How would you relegate it to the new keyword?

You didn't understand what I said: "relegated to **a** new keyword".
I meant something akin to `throws` or `requires`. For example `binds`.
But that's off-topic on this issue. Let's continue our discussion [here](https://github.com/ericelliott/rtype/issues/68#issuecomment-290916496).

This issue is about having a default that permit to omit the definition of an instance object.

**Comment by ericelliott** [comment]
I'm not convinced that a default is needed. Further examples and/or motivation might shed some light on the need for it. Not sure what's wrong with `new Foo() => f` (if you don't care about the specific definition, use a type variable), or `new Foo() => FooInstance` and spell out the instance interface.

**Comment by Mouvedia** [comment]
Since we clearly disagree on the priority of this issue, I am gonna remove this issue from the MVP milestone.

**Comment by ericelliott** [comment]
I think it makes sense that `new Foo()` defaults to `new Foo() => this`, but that doesn't answer the question, what is the interface for `this`?

I think in most cases, you'll want to spell out the instance interface as well, because `new Foo() => this` is essentially equivalent to `new Foo() => Any`, which is equivalent to an untyped return.

**Comment by Mouvedia** [comment]
> `new Foo() => this` is essentially equivalent to `new Foo() => Any`

That's correct; which means that if we had something to omit, it ought to be specific.
In that case the only solution would be to map the constructor's name to its instance's name.

e.g.
```
interface User {
  username: String
}

new User({ username: String })
// would be equivalent to
new User({ username: String }) => User
```

Obviously, unions cannot be covered:
```
new Panthera({ species: String }) => Onca | Leo | Pardus | Uncia | Tigris
```

Would declaring the interface be mandatory?
e.g. we know that `new Foo()` returns Foo by default (even if not declared)

How do you reference a constructor function? Why would you?

**Comment by ericelliott** [comment]
This example:

```js
interface User {
  username: String
}

new User({ username: String })
// would be equivalent to
new User({ username: String }) => User
```

Looks like you're expecting constructor functions in JS to behave like classes from other languages.

A constructor function signature represents the signature for the constructor function, **not** the interface for an instance created by the function. Likewise, because the signature syntax in rtype and the `instance` syntax are equivalent, the example you give here doesn't work. You've essentially declared two interfaces with the same name.

The constructor signature and the instance signature can't be the same in JS because attaching static methods to a constructor in JS is a common idiom, and you'd be conflating static methods with instance methods.

**Comment by Mouvedia** [comment]
This example was taken from https://github.com/ericelliott/rtype#constructors

> because the signature syntax in rtype and the instance syntax are equivalent

Are you talking about the signature syntax of constructor functions?

> How do you reference a constructor function? Why would you?

Could you answer these questions?

TL;DR: if you never reuse/refer that constructor's name there's no conflicts with interfaces' names.

**Comment by ericelliott** [comment]
If that's from the readme, it's a bug that needs fixing. The object returned should be a `UserInstance`.

> TL;DR: if you never reuse/refer that constructor's name there's no conflicts with interfaces' names.

You might, to refer to static props/methods. Because all JS funcs are also objects which can have their own props/methods different from instance props/methods (yet still valuable), we can't safely reuse the names of constructor types and instance types.

Because the function syntax and interface syntax share the same scope, we need to disambiguate.

**Comment by Mouvedia** [comment]
> to refer to static props/methods

Right.
There are PROs and CONs to that. Clearly separating static and instance methods/properties may be useful but having them all on the same interface is convenient too.

There's a convention in the JS community (I don't know if it's shared by other languages) which use `#` for instance m/p:

e.g.
```
interface Array {
  isArray: Function,
  #join: Function,
  …
}

**Comment by ericelliott** [comment]
There are proposals for class static props in JS which may introduce new syntax. Let's wait for that to land before we make a decision on adding syntax (so we can use idiomatic syntax). In the meantime, this is a bug that needs fixing:

```js
interface User {
  username: String
}

new User({ username: String })
// would be equivalent to
new User({ username: String }) => User
```

**Comment by Mouvedia** [comment]
> In the meantime, this is a bug that needs fixing

Go read the chapter: that's not the current example.

> There are proposals for class static props in JS

Are you talking about https://github.com/tc39/proposal-static-class-features?

```
interface ArrayInstance {
  static isArray: Function,
  join: Function,
  …
}

**Comment by ericelliott** [comment]
Yes. Note that the `#` syntax is only for private fields, which aren't useful for type information.

I see that the `static` keyword has been promoted to stage 3. I think that's stable enough to introduce to this spec, if you'd like to PR. =)

**Comment by Mouvedia** [comment]
This issue has been sidetracked.
